egress/src/reports/windrose.rs

```diff
@@ -0,0 +1,1114 @@
+    /// 1D histogram values
+    pub sums: Vec<f64>,
+}
+
+/// Response from reports/windrose/{station_id} endpoint
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct WindroseResp {
+    pub wind_speed: Axis,
+    pub wind_direction: Axis,
+    pub extras: WindCategories,
+    pub table: Vec<Vec<f64>>,
+    pub metadata: Metadata,
+}
+
+pub async fn windrose_handler(
+    Path(station_id): Path<i32>,
+    Query(params): Query<WindroseParams>,
+    State(pools): State<DbPools>,
+    State(tables): State<PatchworkTables>,
+    Extension(roles): Extension<Option<Vec<i32>>>,
+) -> Result<Json<WindroseResp>, (StatusCode, String)> {
+    let roles = roles.unwrap_or_default();
+
+    // NOTE: given how permits work at the moment, open and restricted are mutually exclusive
+    let (open_data, restricted_data) = tokio::try_join!(
+        fetch_wind_data(station_id, &params, &roles, pools.open, tables.open),
+        fetch_wind_data(
+            station_id,
+            &params,
+            &roles,
+            pools.restricted,
+            tables.restricted
+        ),
+    )
+    .map_err(internal_error)?;
+
+    let days = match (open_data.is_empty(), restricted_data.is_empty()) {
+        (false, _) => open_data,
+        (_, false) => restricted_data,
+        (true, true) => {
+            return Err((
+                StatusCode::NOT_FOUND,
+                "no data found for this station".to_string(),
+            ))
+        }
+    };
+
+    // TODO: spawn sync thread here?
+    let windrose = Windrose::new_from_days(SPEED_AXIS, DIRECTION_AXIS, days);
+
+    let metadata = Metadata {
+        station_id,
+        fromtime: params.fromtime,
+        totime: params.totime,
+        number_of_values: windrose.total_obs,
+        months: params.months,
+    };
+
+    Ok(Json(WindroseResp {
+        wind_direction: Axis {
+            labels: WIND_DIRECTION_LABELS,
+            sums: windrose.direction_hist,
+        },
+        wind_speed: Axis {
+            labels: WIND_SPEED_LABELS,
+            sums: windrose.speed_hist,
+        },
+        metadata,
+        extras: windrose.wind_categories,
+        table: windrose.hist,
+    }))
+}
+
+#[derive(Debug, Serialize)]
+pub struct WindroseAvailable {
+    pub station_id: i32,
+    permit: i32,
+    from: DateTime<Utc>,
+    to: Option<DateTime<Utc>>,
+}
+
+#[derive(Debug, Serialize)]
+pub struct WindroseAvailabilityResp {
+    stations: Vec<WindroseAvailable>,
+}
+
+fn is_wind_speed_timeseries(label: &PatchworkLabel) -> bool {
+    label.param_id == WIND_SPEED_PARAM_ID
+        && label.level == DEFAULT_LEVEL
+        && label.sensor == DEFAULT_SENSOR
+}
+
+pub async fn windrose_availability_handler(
+    State(tables): State<PatchworkTables>,
+    Extension(roles): Extension<Option<Vec<i32>>>,
+) -> Result<Json<WindroseAvailabilityResp>, (StatusCode, String)> {
+    // TODO: not sure how performant this is, maybe we need a different data structure?
```

Comment:
You can benchmark it and know for sure

Reply:
True, but needs a bit of refactoring that I'm not willing to spend time on right now. I've remove the TODO